Add additional checks to ext2 loader

Signed-off-by: Frediano Ziglio <frediano.ziglio@citrix.com>

diff --git a/core/fs/ext2/ext2.c b/core/fs/ext2/ext2.c
index 7988faa..cf1c754 100644
--- a/core/fs/ext2/ext2.c
+++ b/core/fs/ext2/ext2.c
@@ -139,6 +139,8 @@ ext2_get_inode(struct fs_info *fs, int inr)
     block_off = inode_offset % EXT2_INODES_PER_BLOCK(fs);
 
     data = get_cache(fs->fs_dev, block_num);
+    if (!data)
+	return NULL;
 
     return (const struct ext2_inode *)
 	(data + block_off * EXT2_SB(fs)->s_inode_size);
