fix typo in comment

From: Frediano Ziglio <frediano.ziglio@citrix.com>

diff --git a/com32/mboot/mboot.c b/com32/mboot/mboot.c
index 35450e0..10e6701 100644
--- a/com32/mboot/mboot.c
+++ b/com32/mboot/mboot.c
@@ -198,7 +198,7 @@ int main(int argc, char *argv[])
     }
 
     if (init_map())
-	return 1;		/* Failed to allocate intitial map */
+	return 1;		/* Failed to allocate initial map */
 
     /*
      * Map the primary image.  This should be done before mapping anything
