diff --git a/core/adv.inc b/core/adv.inc
index 88ea906..b2e3421 100644
--- a/core/adv.inc
+++ b/core/adv.inc
@@ -373,41 +373,72 @@ adv_read_write:
 		push si
 		jmp si
 
+		extern xfer_buf_seg
 .ebios:
+		pusha
+		sub sp,1ch
+		push 1eh
+		mov si,sp
+		mov ah,48h			; EDD get drive parameters
+		mov dl,[ADVDrive]
+		push ds
+		push ss
+		pop ds
+		mov word [si+18h], 512
+		int 13h
+		mov ax, [si+18h]
+		pop ds
+		mov [ADVSsz], ax
+		add sp,1eh
+		popa
+
 		mov cx,adv_retries
 .eb_retry:
 		; Form DAPA on stack
-		cld
-
 		pushad
+		push edx
+		push eax
+		push xfer_buf_seg
+		push 0
+		push word 1			; Sector count
+		push word 16			; DAPA size
 
-		; copy to sector
-		push ax
+		; check if we must do a RMW operation
+		cmp word [ADVSsz], 512
+		je .no_rmw
+		cmp byte [ADVOp], 2
+		je .no_rmw
+
+		mov si,sp
+		mov dl,[ADVDrive]
+		mov ax,4200h
 		push ds
-		push es
-		mov si, bx
-		push es
+		push ss
+		pop ds
+		int 13h
 		pop ds
-		push 2000h
+		jc .eb_error_stack
+
+.no_rmw:
+		; modify sector with ADV
+		cld
+		push es
+		push xfer_buf_seg
 		pop es
 		push di
 		xor di, di
-		mov cx, 256 * 8
+		mov cx, [ADVSsz]
+		shr cx, 1
 		xor ax, ax
 		rep stosw
 		pop di
+		push di
+		mov si, bx
 		mov cx, 256
 		rep movsw
+		pop di
 		pop es
-		pop ds
-		pop ax
 
-		push edx
-		push eax
-		push 2000h
-		push 0
-		push word 1			; Sector count
-		push word 16			; DAPA size
 		mov si,sp
 		mov dl,[ADVDrive]
 		mov ax,4000h
@@ -417,6 +448,7 @@ adv_read_write:
 		pop ds
 		int 13h
 		pop ds
+.eb_error_stack:
 		lea sp,[si+16]			; Remove DAPA
 
 		popad
@@ -427,14 +459,13 @@ adv_read_write:
 		push ds
 		mov si, di
 		mov di, bx
-		push 2000h
+		push xfer_buf_seg
 		pop ds
 		mov cx, 256
 		rep movsw
 		pop ds
 		popa
 
-
 		pop si
 		ret
 .eb_error:
@@ -545,6 +576,7 @@ ADVOff0		dw 0
 ADVOff1		dw 0
 ADVDrive	db -1			; No ADV defined
 ADVCHSInfo	db -1			; We have CHS info for this drive
+ADVSsz		dw 512			; sector size
 
 		section .bss16
 ADVOp		resb 1
